src/java/org/apache/cassandra/db/streaming/CassandraBlockStreamReader.java

```diff
@@ -0,0 +1,177 @@
+        this.format = streamHeader.format;
+        this.fileSeqNum = header.sequenceNumber;
+        this.version = streamHeader.version;
+        this.firstKey = streamHeader.firstKey;
+    }
+
+    /**
+     * @param inputPlus where this reads data from
+     * @return SSTable transferred
+     * @throws IOException if reading the remote sstable fails. Will throw an RTE if local write fails.
+     */
+    @SuppressWarnings("resource") // input needs to remain open, streams on top of it can't be closed
+    @Override
+    public SSTableMultiWriter read(DataInputPlus inputPlus) throws IOException
+    {
+        long totalSize = manifest.getTotalSize();
+
+        ColumnFamilyStore cfs = ColumnFamilyStore.getIfExists(tableId);
+
+        if (cfs == null)
+        {
+            // schema was dropped during streaming
+            throw new IOException("CF " + tableId + " was dropped during streaming");
+        }
+
+        logger.debug("[Stream #{}] Start receiving file #{} from {}, size = {}, ks = '{}', table = '{}'.",
+                     session.planId(), fileSeqNum, session.peer, totalSize, cfs.keyspace.getName(), cfs.getTableName());
+
+        BigTableBlockWriter writer = null;
+
+        try
+        {
+            writer = createWriter(cfs, totalSize, manifest.getComponents());
+            long bytesRead = 0;
+            for (Component component : manifest.getComponents())
+            {
+                long length = manifest.getSizeForType(component.type);
+
+                logger.debug("[Stream #{}] About to receive file {} from {} readBytes = {}, componentSize = {}, totalSize = {}",
+                             session.planId(), component, session.peer, FBUtilities.prettyPrintMemory(bytesRead),
+                             FBUtilities.prettyPrintMemory(length), FBUtilities.prettyPrintMemory(totalSize));
+
+                writer.writeComponent(component.type, inputPlus, length);
+                session.progress(writer.descriptor.filenameFor(component), ProgressInfo.Direction.IN, length, length);
+                bytesRead += length;
+
+                logger.debug("[Stream #{}] Finished receiving file {} from {} readBytes = {}, componentSize = {}, totalSize = {}",
+                             session.planId(), component, session.peer, FBUtilities.prettyPrintMemory(bytesRead),
+                             FBUtilities.prettyPrintMemory(length), FBUtilities.prettyPrintMemory(totalSize));
+            }
+
+            writer.descriptor.getMetadataSerializer().mutateLevel(writer.descriptor, sstableLevel);
+            return writer;
+        }
+        catch (Throwable e)
+        {
+            logger.error("[Stream {}] Error while reading from stream on ks='{}' and table='{}'.",
+                        session.planId(), cfs.keyspace.getName(), cfs.getTableName(), e);
+            if (writer != null)
+            {
+                writer.abort(e);
+            }
+            throw Throwables.propagate(e);
```

Comment:
`Throwables.propagate()` is deprecated now, so we should probably not be using it in new code. Guava developers on why - [here](https://github.com/google/guava/wiki/Why-we-deprecated-Throwables.propagate).

Reply:
Cherry picked. Appreciate the link.